Add initial docs for Go gRPC. (#2)
smallstep/practicalzt: src/docs/go-grpc/sections/03-operationalize/01-renewal/01-builtin.mdx

The above example will work as long as the certificates do not expire. By

> default, `step-ca` certificates expire after 24 hours, so to go above that, we

Reviewer: Do we want to call them `step-ca` certificates in PZT, or Certificate Manager certificates, or something else?
@alanchrt @mikemaxey

Reviewer 2: Leave it without the modifier. Just "certificates"

Reviewer 2: Or. Use. "smallstep certificates" expire...

Author: Fixed in 6022ce7

need a way to renew the server certificates automatically.

One of the best options is to use the ACME protocol, and instead of using `step`

> to create the certificate, we will use the package

Reviewer: ```suggestion
to create the certificate, we will use the Go package
```

[acme/autocert](https://pkg.go.dev/golang.org/x/crypto/acme/autocert) that
provides an easy way to support ACME on Go gRPC (and HTTP) servers. It provides
automatic access to certificates from Let's Encrypt or a private ACME server
like `step-ca`.

> Using `acme/autocert` the `autocert.Manager` will now take care of creating a

Reviewer: ```suggestion
Using `acme/autocert`, the `autocert.Manager` will now take care of creating a
```

certificate signing request (CSR) and requesting a certificate using the ACME
protocol. It will also take care of renewing the certificate, to do this we need
to make sure to set a proper `RenewBefore` according to the lifetime of your

> certificates. By default, `RenewBefore` is set to 30 days before the expiration,
> but if our certificate lifetime is 24 hours the renewal will kick on every
> connection, to avoid this we can for example renew 8 hours before the expiration
> setting `RenewBefore` to `8 * time.Hour`.

Reviewer: ```suggestion
certificates. By default, `RenewBefore` is set to 30 days before the expiration.
Since our certificate lifetime is 24 hours, the `autocert.Manager` will attempt renewal _on every connection_ unless we adjust this setting.
We recommend attempting renewal when a certificate has exceeded two-thirds of its lifetime.
To attempt renewal 8 hours before the certificate expires, set `RenewBefore` to `8 * time.Hour`.
```

> Here is the main function using `acme/autocert`:

Reviewer: ```suggestion
Putting all the pieces together,
here is the updated `main` function for our server, with `acme/autocert` integrated into it:
```

transport.TLSClientConfig = &tls.Config{
	RootCAs: pool,
}

// Create the acme client configured with our private ACME server
client = &acme.Client{
	Key:          key,
	DirectoryURL: directory,
	HTTPClient:   &http.Client{
		Transport: transport,
	},
}

// Configure the acme/autocert manager that will get the server certificate
// using the tls-alpn-01 challenge.
m := &autocert.Manager{
		Prompt:      autocert.AcceptTOS,
		HostPolicy:  autocert.HostWhitelist(domain),
		Client:      client,
		Cache:       autocert.DirCache("secret-dir"),
		RenewBefore: 8 * time.Hour,
}

// Prepare a listener in the port :443, required for the tls-alpn-01 challenge
lis, err := net.Listen("tcp", ":443")
if err != nil {
	log.Fatalf("failed to listen: %v", err)
}

// Start the gRPC server with the TLS credentials from the manager
srv := grpc.NewServer(grpc.Creds(credentials.NewTLS(m.TLSConfig())))
helloworld.RegisterGreeterServer(srv, &helloServer{})
reflection.Register(srv)

log.Printf("server listening at %v", lis.Addr())
if err := srv.Serve(lis); err != nil {
	log.Fatalf("failed to serve: %v", err)
}
```

The CSRs created by `acme/autocert` will use a ECDSA P-256 key, but if a client

> does not support ii, it will automatically create a CSR with an RSA key size of

Reviewer: ```suggestion
does not support it, it will automatically create a CSR with an RSA key size of
```

> 2048 bits and do the ACME flow to get it sign.

Reviewer: ```suggestion
2048 bits and do the ACME flow to get it signed.
```

TLS cipher suites and the TLS version can be configured in the manager's

> `*tls.Config`. The default behavior is to use a safe list of all the ones

Reviewer: ```suggestion
`*tls.Config`. The default behavior is to use a safe list of all the cipher suites
```
